docs/rules/jsx-curly-newline.md

```diff
@@ -0,0 +1,125 @@
+# Enforce linebreaks in curly braces in JSX attributes and expressions. (react/jsx-curly-newline)
+
+Many style guides require or disallow newlines inside of jsx curly expressions.
+
+**Fixable:** This rule is automatically fixable using the `--fix` flag on the command line.
+
+## Rule Details
+
+This rule enforces consistent linebreaks inside of curlies of jsx curly expressions.
+
+## Rule Options
+
+This rule has a string option:
+
+* `consistent` enforces either both of linebreaks around curlies are present, or none are present.
+* `multiline` enforces that when the contained expression spans multiple line, require both linebreaks. When the contained expression is single line, disallow both line breaks.
+* `multiline-lax` (default) enforces that when the contained expression spans multiple line, require both linebreaks. When the contained expression is single line, disallow inconsistent line break.
```

Comment:
this seems like `consistent-multiline` would be a better name

Reply:
I think that consistent multiline sound like a consistent version of multiline, but that is not true. Because multipline is already consistent.

Reply:
I think multilinelax is fit because it is like multiline, but more permissive when the expression is single line.

Reply:
Looking at eslint core, there's a number of rules that have an object option with a `multiline` boolean, and a `consistent` boolean - if we only allowed multiline, multiline + consistent, and consistent, that seems like it'd cover it more clearly?

Reply:
`multiline + consistent = multiline`. `multiline` already implied consistent.

Reply:
in `multiline` either both newline are present, or both newline are absent.  So a code that is valid in `multiline` is also valid in `consistent`. Consistent is the "weakest" option.

Reply:
I don't find the current schema clear.

It sounds like you want the following characteristics controllable:
1. force both linebreaks on a multiline expression, or not
1. disallow inconsistent linebreaks on a one-line expression, or disallow any linebreaks on a one-line expression.

So, what about:
```js
"multiline": "consistent" // or "ignore" or "forbid" or "require"
"single-line": "consistent" // or "ignore" or "forbid" or "require"
```
That way any setting can be configured without a subjective word like "lax", and it's extensible since it's an object option.

Reply:
This might be good. I was trying to think in the box of precedent eslint core rules, and tries to follow the convention established there.

Reply:
I don't like the option schema
> ```js
> "multiline": "consistent" // or "ignore" or "forbid" or "require"
> "single-line": "consistent" // or "ignore" or "forbid" or "require"
> ```
because there are 4*4 combinations, therefore many test cases to write. But in practice only a few of them make sense, so instead might as well just pick the few good options and give them a string name like `multiline`. It also saves users time by reducing their choices.

Reply:
Which don't make sense? We can use schema validation to only permit the desired combinations.

Reply:
Now I feel I was wrong thinking most of them don't make sense. However I think it is safe to not support the `ignore` option (just use `consistent`).

Reply:
How about this option schema?
```ts
"never" |
"consistent" |
{
  "single-line": "require" | "forbid" | "consistent" ,
  "multiline": "require" | "forbid" | "consistent" ,
}
```

Reply:
That seems good; let's update the PR and see what the test cases and docs for each combination look like :-D

Reply:
Now I feel I am ok to live with the simple option schema `"never" | "consistent"`, dropping the multiline option. Are you interested in this direction?